scripts/ci/dag.py

```diff
@@ -0,0 +1,169 @@
+from __future__ import annotations
+
+import time
+from concurrent.futures import ThreadPoolExecutor
+from multiprocessing import Event, cpu_count
+from multiprocessing.synchronize import Event as EventClass
+from queue import Empty, Queue
+from typing import Callable, Generic, Hashable, TypeVar
+
+_T = TypeVar("_T", bound=Hashable)
+
+
+class _Node(Generic[_T]):
+    def __init__(self, value: _T):
+        self.value = value
+        self.counter: int = 0
+        self.dependents: list[_Node[_T]] = []
+
+
+class DAG(Generic[_T]):
+    def __init__(self, dependency_graph: dict[_T, list[_T]]):
+        """
+        Construct a directed acyclic graph from an adjacency list.
+
+        The `dependency_graph` _must not_ contain any cycles.
+        """
+
+        self._nodes: dict[_T, _Node[_T]] = {}
+        self._queue: list[_T] = []
+        self._num_finished: int = 0
+
+        for node, deps in dependency_graph.items():
+            new_node = self._get_or_insert(node)
+            new_node.counter += len(deps)
+            for dep in deps:
+                self._get_or_insert(dep).dependents.append(new_node)
+
+        self._queue.extend(node.value for node in self._nodes.values() if node.counter == 0)
+
+    def _get_or_insert(self, node: _T) -> _Node[_T]:
+        if node not in self._nodes:
+            self._nodes[node] = _Node(node)
+        return self._nodes[node]
```

Comment:
```suggestion
    def _get_or_insert(self, value: _T) -> _Node[_T]:
        if value not in self._nodes:
            self._nodes[value] = _Node(value)
        return self._nodes[value]
```

Reply:
actually, `key` is probably a better name, considering how it is used

Reply:
I think `node` is a good name, because it is a node in the graph. The specific graph representation here just happens to use a dictionary. `_Node` is imho more deserving of a rename to `_NodeState`, and `_nodes` to `_node_states`.

Reply:
sure, as long as it is consistent, and the same name isn't reused!